unified-runtime/source/adapters/level_zero/v2/queue_immediate_out_of_order.cpp

```diff
@@ -0,0 +1,182 @@
+  }
+
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t ur_queue_immediate_out_of_order_t::queueGetNativeHandle(
+    ur_queue_native_desc_t * /*pDesc*/, ur_native_handle_t *phNativeQueue) {
+  *phNativeQueue = reinterpret_cast<ur_native_handle_t>(
+      (*commandListManagers.get_no_lock())[getNextCommandListId()]
+          .getZeCommandList());
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t ur_queue_immediate_out_of_order_t::queueFinish() {
+  TRACK_SCOPE_LATENCY("ur_queue_immediate_out_of_order_t::queueFinish");
+
+  auto commandListManagersLocked = commandListManagers.lock();
+
+  for (size_t i = 0; i < numCommandLists; i++) {
+    ZE2UR_CALL(zeCommandListHostSynchronize,
+               (commandListManagersLocked[i].getZeCommandList(), UINT64_MAX));
+    UR_CALL(commandListManagersLocked[i].releaseSubmittedKernels());
+  }
+
+  hContext->getAsyncPool()->cleanupPoolsForQueue(this);
+  hContext->forEachUsmPool([this](ur_usm_pool_handle_t hPool) {
+    hPool->cleanupPoolsForQueue(this);
+    return true;
+  });
+
+  return UR_RESULT_SUCCESS;
+}
+
+ur_result_t ur_queue_immediate_out_of_order_t::queueFlush() {
+  return UR_RESULT_SUCCESS;
+}
+
+ur_queue_immediate_out_of_order_t::~ur_queue_immediate_out_of_order_t() {
+  try {
+    UR_CALL_THROWS(queueFinish());
+
+    for (size_t i = 0; i < numCommandLists; i++) {
+      barrierEvents[i]->release();
+    }
+
+    ur::level_zero::urContextRelease(hContext);
+  } catch (...) {
+    // Ignore errors during destruction
+  }
+}
+
+ur_result_t ur_queue_immediate_out_of_order_t::enqueueEventsWaitWithBarrier(
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
+  TRACK_SCOPE_LATENCY(
+      "ur_queue_immediate_out_of_order_t::enqueueEventsWaitWithBarrier");
+  // Since we use L0 in-order command lists, we don't need a real L0 barrier,
+  // just wait for requested events in potentially different queues and add a
+  // "barrier" event signal because it is already guaranteed that previous
+  // commands in this queue are completed when the signal is started. However,
+  // we do need to use barrier if profiling is enabled: see
+  // zeCommandListAppendWaitOnEvents
+  bool needsRealBarrier = (flags & UR_QUEUE_FLAG_PROFILING_ENABLE) != 0;
+  auto appendEventsWaitFn =
+      needsRealBarrier ? &ur_command_list_manager::appendEventsWaitWithBarrier
+                       : &ur_command_list_manager::appendEventsWait;
+
+  auto commandListManagersLocked = commandListManagers.lock();
+
+  // Enqueue wait for the user-provider events on the first command list.
+  std::invoke(appendEventsWaitFn, commandListManagersLocked[0],
+              numEventsInWaitList, phEventWaitList, barrierEvents[0]);
+
+  // Submit barrier or request barrierEvents[id] to be signaled on remaining
+  // command lists.
+  for (size_t id = 1; id < numCommandLists; id++) {
+    std::invoke(appendEventsWaitFn, commandListManagersLocked[id], 0, nullptr,
```

Comment:
will this actually be functionally equivalent to a barrier on a single queue? Don't we need a single event that the command lists need to converge on?

In other words, in this scenario:
enqueueA(no events); -> this goes to queue 0
barrier(no output event);
enqueueB(no events); -> this goes to queue 1

will this barrier implementation actually prevent enqueueB from proceeding before enqueueA finishes? If so, why?